glp/neighborlist.py

```diff
@@ -189,23 +212,210 @@ def candidates_fn(n):
+        scatter_indices = jnp.reshape(bin_id, (-1,))
+        bin_value = jnp.reshape(bin_value, (-1,) + bin_value.shape[-2:])
+        return value.at[scatter_indices].set(bin_value)
+    neighbor_idx = jnp.zeros((N + 1,) + bin_idx.shape[-2:], jnp.int32)
+    neighbor_idx = copy_values_from_bin(neighbor_idx, bin_idx, idx)
+    # Reshape the neighbor_idx to get the centers and others
+    others = jnp.reshape(neighbor_idx[:-1, :, 0], (-1,))
+    centers = jnp.repeat(jnp.arange(0, N), others.shape[0] // N)
+    return centers, others
 
-def get_neighbors(positions, cell, square_distances, cutoff, padding_mask=None):
-    centers, others = candidates_fn(positions.shape[0])
-    sq_distances = square_distances(positions[centers], positions[others])
-
-    mask = sq_distances <= cutoff
-    mask = mask * (centers != others)  # remove self-interactions
+def get_neighbors(positions, square_distances, cutoff, padding_mask=None, cl=None):
+    N, dim = positions.shape
+    if cl is not None:
+        centers, others = cell_list_candidate_fn(cl.id, N, dim)
+    else:
+        centers, others = candidates_fn(N)
 
-    # mask out fake positions
+    
+    sq_distances = square_distances(positions[centers], positions[others])
+    mask = sq_distances <= (cutoff**2)
+    mask = mask * ((centers != others) & (others<N))  # remove self-interactions and neighbors repetitions
     if padding_mask is not None:
         mask = mask * padding_mask[centers]
         mask = mask * padding_mask[others]
 
     hits = jnp.sum(mask)
-
     return centers, others, sq_distances, mask, hits
 
 
 def make_squared_distance(cell):
     return vmap(lambda Ra, Rb: squared_distance(displacement(cell, Ra, Rb)))
+
+
+def cell_list(cell, cutoff, buffer_size_multiplier=1.25, bin_size_multiplier=1):
+    """Implementation of cell list neighborlist in pbc."""
+
+    cell = jnp.array(cell)
+    cutoff *= bin_size_multiplier
+    def allocate_fn(positions, extra_capacity=0):
+        # This function is not jittable, we're determining shapes
+        N = positions.shape[0]
+        _, bin_size, bins_per_side, bin_count = bin_dimensions(cell, cutoff)
+        bin_capacity = estimate_bin_capacity(positions, cell, bin_size,
+                                            buffer_size_multiplier)
+        # Computing the maximum number of atoms per bin (capacity)
+        # extra_capacity is used to increase the capacity of the bins to avoid reallocation
+        bin_capacity += extra_capacity
+        
+        overflow = False
+        bin_id = N * jnp.ones((bin_count * bin_capacity, 1), dtype=jnp.int32)
+
+        # Compute the hash of each particle to assign it to a bin
+        hash_multipliers = compute_hash_constants(bins_per_side)
+        particle_id = iota(jnp.int32, N)
+        indices = jnp.array(jnp.floor(positions @ inverse(bin_size).T), dtype=jnp.int32)
+        # Some particles are in the edge and might have negative indices or larger than bins_per_side
```

Comment:
I have tried wrapping the positions instead of the indices. However, when I did the wrap I had negative values (probably due to the number of decimals) which made it have negative indices, which broke the implementation. So, the simplest way was just to wrap the indices having as 'cell vectors' the `bins_per_side`.